### PR TITLE
Hash based config entry replication

### PR DESCRIPTION
**Pull request by dhiaayachi**
### Description

This is another take on #19734, This version calculate a deterministic hash based on the config entry content, that way we can guarantee that the hash will not vary from one server to another in case of snapshot restore.

### Testing & Reproduction steps
- added test to diffConfig
- changed tests to account for hash

### PR Checklist

* [ ] updated test coverage
* [ ] external facing docs updated
* [ ] appropriate backport labels added
* [ ] not a security concern

### CODE REVIEW

**Review by rboyer — APPROVED** [review]

File: proto/private/pbconfigentry/config_entry.proto [hunk 1]
```diff
@@ -809,8 +815,9 @@ message ResourceReference {
 message BoundAPIGateway {
   map<string, string> Meta = 1;
   repeated BoundAPIGatewayListener Listeners = 2;
+  uint64 Hash = 3;
   // mog: func-to=serviceRefsToStructs func-from=serviceRefFromStructs
-  map<string, ListOfResourceReference> Services = 3;
+  map<string, ListOfResourceReference> Services = 4;
 }
```
Comment (analogue) [inline proto/private/pbconfigentry/config_entry.proto:821]: Is changing the unique field number here for `Services` intentional. Something something backwards compatibility / protobuf best practices?

Comment (dhiaayachi) [reply proto/private/pbconfigentry/config_entry.proto:821]: it's definitely not intentional, I think my IDE is doing things by it self 😅! Great catch!

File: agent/structs/config_entry_inline_certificate.go [hunk 2]
```diff
@@ -31,8 +31,17 @@ type InlineCertificateConfigEntry struct {
...
+	RaftIndex          `hash:"ignore"`
+}
+
+func (e *InlineCertificateConfigEntry) SetHash(h uint64) {
+	e.Hash = h
+}
+
+func (e *InlineCertificateConfigEntry) GetHash() uint64 {
+	return e.Hash
 }
```
[the remark this answers is not in the document]
Comment (dhiaayachi) [reply agent/structs/config_entry_inline_certificate.go]: yes I missed that one, added a test to check that all the config entries have the hash as part of normalize.

**Comment by komapa** [comment]
🙇🏼

**Pull request merged**